mem: Fix "Need is_secure arg" prefetcher crash

erin-le opened the pull request:
This PR fixes the "Need is_secure arg" crash that occurs when using IndirectMemoryPrefetcher, SignaturePathPrefetcher, SignaturePathPrefetcherV2, STeMSPrefetcher, and PIFPrefetcher. This was done by changing some variables to have the type AssociativeSet<...> instead of AssociativeCache<...> and adding in "false" or an existing value for the value of the secure bit in some function calls. Further changes may be needed to move away from hard-coding values.

erin-le committed:
mem: Fix "Need is_secure arg" prefetcher crash

This commit fixes the "Need is_secure arg" crash that occurs when
using the IndirectMemoryPrefetcher, SignaturePathPrefetcher,
SignaturePathPrefetcherV2, STeMSPrefetcher, and PIFPrefetcher. This
was done by changing some variables to be AssociativeSet<...>
instead of AssociativeCache<...> and changing the affected function
calls.

Change-Id: I61808c877514efeb73ad041de273ae386711acae

erin-le committed:
mem: use is_secure instead of hardcoded false in prefetcher crash

This modifies the crash fix so that the function calls that were
modified use a local variables called `is_secure` instead of a
hardcoded `false`. Some of these existed previously so it made
more sense to use them, while others were newly added in to mark
where the code might need to be changed later.

Change-Id: I0c0d14b74f0ccf70ee5fe7c8b01ed0266353b3c1

giactra commented:
Yeah, I have a patchset which is fixing this by allowing the AssociativeCache to use generic tags... Anyway it makes sense to patch this for now.. This should definitely be backported into v24.0 I believe @BobbyRBruce

hnpl commented on src/mem/cache/prefetch/associative_set.hh:
```diff
@@ -90,7 +90,7 @@ class AssociativeSet : public AssociativeCache<Entry>
     using AssociativeCache<Entry>::findEntry;
     using AssociativeCache<Entry>::insertEntry;
-    using AssociativeCache<Entry>::getPossibleEntries;
+    // using AssociativeCache<Entry>::getPossibleEntries;
```
Can you either remove this line, or provide comments in the code on why this line is commented out?

hnpl commented on src/mem/cache/prefetch/pif.cc:
```diff
@@ -176,14 +176,16 @@ PIF::notifyRetiredInst(const Addr pc)
                 historyBuffer.push_back(spatialCompactor);
-
-                auto idx_entry = index.findEntry(spatialCompactor.trigger);
+                bool is_secure = false;
```
Since the `is_secure` variable is not changed afterwards, can you change this to `constexpr bool is_secure = false;`?

hnpl commented on src/mem/cache/prefetch/signature_path.cc:
```diff
@@ -192,7 +192,8 @@ SignaturePath::getSignatureEntry(Addr ppn, bool is_secure,
 SignaturePath::getPatternEntry(Addr signature)
 {
-    PatternEntry* pattern_entry = patternTable.findEntry(signature);
+    bool is_secure = false;
```
Since the `is_secure` variable is not changed afterwards, can you change this to `constexpr bool is_secure = false;`?

hnpl commented on src/mem/cache/prefetch/signature_path_v2.cc:
```diff
@@ -125,7 +125,8 @@ SignaturePathV2::handlePageCrossingLookahead(signature_t signature,
     assert(gh_entry != nullptr);
     // Any address value works, as it is never used
-    globalHistoryRegister.insertEntry(0, gh_entry);
+    bool is_secure = false;
```
Since the `is_secure` variable is not changed afterwards, can you change this to `constexpr bool is_secure = false;`?

hnpl commented on src/mem/cache/prefetch/signature_path_v2.cc:
```diff
@@ -125,7 +125,8 @@ SignaturePathV2::handlePageCrossingLookahead(signature_t signature,
     // Any address value works, as it is never used
-    globalHistoryRegister.insertEntry(0, gh_entry);
+    bool is_secure = false;
+    globalHistoryRegister.insertEntry(0, is_secure, gh_entry); // false,
```
Can you remove this comment?

hnpl reviewed (commented):
There are a couple of nitpicking comments, but other than that, the code looks good to me.

powerjg commented:
@erin-le can you make the updates suggested by Hoa?

erin-le committed:
mem: Comment removal and adding constexpr to is_secure bools

This commit removes some comments and adds constexpr in front
of "bool is_secure..." in pif.cc, signature_path.cc, and
signature_path_v2.cc

Change-Id: Icafe1d7c97d1d3fbf6abc12ba87ebb596255b96f

odanrc commented:
I don´t really understand what is the bug from the PR description, so I have a couple of questions.

I see that this PR mostly reverts https://github.com/gem5/gem5/pull/745 within the prefetcher realm. Doesn´t the bug also affect src/mem/cache/prefetch/delta_correlating_prediction_tables.hh and other places that use AssociativeCache? Why is this only a partial - although close to full - reversal? i.e., why don´t we fully revert PR 745?

hnpl commented:
The `Need is_secure arg` panic refers to a gem5 panic related to `TaggedEntry`.
https://github.com/gem5/gem5/pull/745/files#diff-eccdc055cc61dd6193ad97da8fa8f08756cf86d59e9d0c3bad8596267741709aR115

I'm not sure if fully reverting #745 for now makes more sense. Given that @giactra already has some patchset for this bug, this PR is more of a temporary fix.

@erin-le Thanks! There is a comment that was not addressed. Also, can you test other prefetchers that were affected by #745, such as `SlimAMPM`?

erin-le committed:
mem: remove stray comment from signature_path_v2.cc

Change-Id: I5ddd2ddd6a9cb4fb032b48870c5ef6b0dc9533c0

erin-le commented:
Sorry about that, it should be removed now. Regarding testing, I went through the following prefetchers

```
MultiPrefetcher StridePrefetcherHashedSetAssociative StridePrefetcher TaggedPrefetcher 
IndirectMemoryPrefetcher SignaturePathPrefetcher SignaturePathPrefetcherV2 
AMPMPrefetcher DCPTPrefetcher IrregularStreamBufferPrefetcher SlimAMPMPrefetcher 
BOPPrefetcher SBOOEPrefetcher STeMSPrefetcher PIFPrefetcher
```
which should be all of the ones in `src/mem/cache/prefetch/Prefetcher.py`.

After the changes, all of them except for StridePrefetcherHashedSetAssociative work. The error message is as follows.

```
TypeError: BasePrefetcher.__init__() takes 1 positional argument but 2 were given
Error setting param L1ICache.prefetcher to <orphan StridePrefetcherHashedSetAssociative>
```
This didn't look like it was related to the crashes that were patched in this PR, so I opened a separate issue for it [here](https://github.com/gem5/gem5/issues/1371).

hnpl commented on src/mem/cache/prefetch/spatio_temporal_memory_streaming.cc:
```diff
@@ -219,9 +222,11 @@ STeMS::reconstructSequence(
 
     // Now query the PST with the PC of each RMOB entry
     idx = 0;
+    bool is_secure = false;
```
Ditto

hnpl reviewed (approved):
Thanks for looking into this @erin-le. There are a couple of nitpicks, but other than that, the code looks good to me!

hnpl commented on src/mem/cache/prefetch/spatio_temporal_memory_streaming.cc:
```diff
@@ -92,12 +92,15 @@ STeMS::checkForActiveGenerationsEnd(const CacheAccessor &cache)
             if (generation_ended) {
                 // PST is indexed using the PC (secure bit is unused)
-                auto pst_entry = patternSequenceTable.findEntry(pst_addr);
+                bool is_secure = false;
```
Can you change this to `constexpr bool is_secure = false;`?

erin-le committed:
mem: Add constexprs to spatio_temporal_memory_streaming.cc

Change-Id: I6fa3d9f9a9d89d59d9ec1fc97c152bea3059f87d

BobbyRBruce commented:
@giactra @erin-le @hnpl : I'm merging this as it Erin appears to have addressed all the concerns and it looks good to me. However, i don't have a good grasp of the importance of this. Is this something that should be hotfixed to stable? It's fixing a bug impacting users? I know Giacomo said this 2 weeks back but just triple checking this is the case.

hnpl commented:
@BobbyRBruce I think this is a candidate for a hotfix. The affected prefetchers are unusable without this fix.

BobbyRBruce commented:
Ok. I'll get that done this week.

BobbyRBruce commented:
Hotfix created here: https://github.com/gem5/gem5/pull/1425
